### DISCUSSION TITLE
尝试用一个自己写的库用于从加密的zip包里加载lua文件

### DISCUSSION THREAD

**Discussion by sue602** [cloudwu/skynet#1599] [Ideas]
1. 如题
2. 没有优雅的解决 skynet.newservice("servicename") 加载问题。
3. 压缩包里的解压操作，如果服务多的情况下，需要重复加载才能解决require路径问题，同样不是很优雅，最好能只在主服务加载一次即可，看例子 https://github.com/sue602/lua-chunk

@cloudwu 云大有没有比较好的思路能否提供一下？

**Comment by cloudwu** [comment]
你可以写一个通用服务，把其它服务名作为这个服务的参数传入。然后用这个通用服务去加载所有你的自定义服务。

service 这个库就是这个做法。

另外，lua  服务的加载器本身就可以在 config 中指定的 https://github.com/cloudwu/skynet/blob/master/service-src/service_snlua.c#L425
只是默认使用  https://github.com/cloudwu/skynet/blob/master/lualib/loader.lua 而已。

**Reply by sue602** [reply]
> 个通用服务去加

require "filename"  如果这个文件是在压缩包里，会导致require加载不到，因为 package.searchpath 找不到
https://github.com/cloudwu/skynet/blob/master/3rd/lua/loadlib.c#L666

**Reply by hanxi** [reply]
参考 service 库搞吧，service 库是使用通用服务 service_provider 来加载一段打包好的代码的。你把代码解压后传进去就行，写个 zipservice 库就行了，做好缓存，第一次解压，后面就不用解压了。

https://github.com/cloudwu/skynet/blob/master/lualib/skynet/service.lua#L29

当然如果你的代码比较复杂，还需要改 loader 的，比如从内存中加载代码，可以参考这个 loader

https://github.com/ers35/luastatic/blob/master/luastatic.lua#L341-L363

更多详细用法还是 RTFM 吧

http://cloudwu.github.io/lua53doc/manual.html#pdf-package.searchers

**Reply by sue602** [reply]
service库可以解决一部分需求，但是未达到优雅的从加密的压缩包中加载lua代码（源码不泄漏）。
@hanxi

**Reply by hanxi** [reply]
一般都是改 luavm 就够了，加个密码或者 xor 或者改一下字节码的顺序，luac 跑一下就行了，发布时换个加密后的 loader。压缩包还是没必要吧，解压是需要密码的，密码在内存里查一下就出来了，改字节码就加大破解难度了。

**Reply by sue602** [reply]
我们是已经完成skynet项目工程的lua源码加密、解密方案，这个方案是无侵入式的。
现在是有想针对开发过程，一些公司核心项目核心代码的加密，以防核心代码泄漏。
@hanxi

**Reply by Gowa2017** [reply]
luac 不行么
